File: backend/registrations/walkInRegistration.mjs
import { PutCommand, GetCommand, QueryCommand } from "@aws-sdk/lib-dynamodb";
import { db, REGISTRATIONS_TABLE, EVENTS_TABLE } from "./shared/db.mjs";
import { ok, badRequest, conflict, serverError, cors } from "./shared/response.mjs";
import { newId, registrationNumber } from "./shared/ids.mjs";
import { audit, callerFromEvent } from "./shared/auth.mjs";

export async function handler(event) {
  if (event.httpMethod === "OPTIONS") return cors();

  const caller = callerFromEvent(event);
  const { eventId } = event.pathParameters ?? {};
  if (!eventId) return badRequest("eventId is required");

  let body;
  try {
    body = JSON.parse(event.body ?? "{}");
  } catch {
    return badRequest("Invalid JSON");
  }

  const { fullName, organisation, email, phone, position } = body;
  if (!fullName || fullName.trim().length < 2) return badRequest("fullName is required");
  if (!organisation || organisation.trim().length < 2) return badRequest("organisation is required");
  if (!email || !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email.trim())) return badRequest("Valid email is required");

  const eventResult = await db.send(new GetCommand({ TableName: EVENTS_TABLE, Key: { eventId } }));
  if (!eventResult.Item) return badRequest("Event not found");

  // Duplicate check
  const existing = await db.send(new QueryCommand({
    TableName: REGISTRATIONS_TABLE,
    IndexName: "email-eventId-index",
    KeyConditionExpression: "email = :email AND eventId = :eventId",
    ExpressionAttributeValues: { ":email": email.toLowerCase().trim(), ":eventId": eventId },
    Limit: 1,
  }));
  if ((existing.Items ?? []).length > 0) return conflict("This email is already registered for this event");

  const item = {
    registrationId: newId(),
    eventId,
    registrationNumber: registrationNumber(
      eventResult.Item.name.replace(/\s+/g, "").slice(0, 6).toUpperCase()
    ),
    fullName: fullName.trim(),
    organisation: organisation.trim(),
    email: email.toLowerCase().trim(),
    phone: phone?.trim() ?? null,
    position: position?.trim() ?? null,
    registrationType: "walk_in",
    checkedInAt: null,
    checkedInBy: null,
    badgePrintedAt: null,
    badgePrintCount: 0,
    createdBy: caller.actorId ?? null,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
  };

  try {
    await db.send(new PutCommand({ TableName: REGISTRATIONS_TABLE, Item: item }));
    await audit("participant.walk_in_registered", {
      entity: "participant",
      entityId: item.registrationId,
      actorId: caller.actorId,
      actorLabel: caller.actorLabel,
      meta: { name: item.fullName, email: item.email, phone: item.phone, reg: item.registrationNumber },
    });
    return ok(item);
  } catch (e) {
    console.error(e);
    return serverError();
  }
}

File: backend/registrations/walkInRegistration.py
import json
import logging
import re

from shared.auth import audit, caller_from_event, iso_now
from shared.db import events_table, registrations_table
from shared.ids import new_id, registration_number
from shared.response import bad_request, conflict, cors, ok, server_error

log = logging.getLogger()

EMAIL_RE = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")


def handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return cors()

    caller = caller_from_event(event)
    event_id = (event.get("pathParameters") or {}).get("eventId")
    if not event_id:
        return bad_request("eventId is required")

    try:
        body = json.loads(event.get("body") or "{}")
    except (ValueError, TypeError):
        return bad_request("Invalid JSON")

    full_name = body.get("fullName")
    organisation = body.get("organisation")
    email = body.get("email")
    phone = body.get("phone")
    position = body.get("position")

    if not full_name or len(full_name.strip()) < 2:
        return bad_request("fullName is required")
    if not organisation or len(organisation.strip()) < 2:
        return bad_request("organisation is required")
    if not email or not EMAIL_RE.match(email.strip()):
        return bad_request("Valid email is required")

    email_norm = email.strip().lower()

    event_item = events_table().get_item(Key={"eventId": event_id}).get("Item")
    if not event_item:
        return bad_request("Event not found")

    existing = registrations_table().query(
        IndexName="email-eventId-index",
        KeyConditionExpression="email = :email AND eventId = :eventId",
        ExpressionAttributeValues={":email": email_norm, ":eventId": event_id},
        Limit=1,
    )
    if existing.get("Items"):
        return conflict("This email is already registered for this event")

    prefix = re.sub(r"\s+", "", event_item.get("name") or "")[:6].upper()
    now = iso_now()
    item = {
        "registrationId": new_id(),
        "eventId": event_id,
        "registrationNumber": registration_number(prefix),
        "fullName": full_name.strip(),
        "organisation": organisation.strip(),
        "email": email_norm,
        "phone": phone.strip() if phone else None,
        "position": position.strip() if position else None,
        "registrationType": "walk_in",
        "checkedInAt": None,
        "checkedInBy": None,
        "badgePrintedAt": None,
        "badgePrintCount": 0,
        "createdBy": caller["actorId"],
        "createdAt": now,
        "updatedAt": now,
    }

    try:
        registrations_table().put_item(Item=item)
        audit("participant.walk_in_registered", entity="participant",
              entity_id=item["registrationId"],
              actor_id=caller["actorId"], actor_label=caller["actorLabel"],
              meta={"name": item["fullName"], "email": item["email"],
                    "phone": item["phone"], "reg": item["registrationNumber"]})
        return ok(item)
    except Exception:
        log.exception("walkInRegistration failed")
        return server_error()
